### ISSUE TITLE
applying affine is creating some distortion or an illusion

### ISSUE THREAD

**Issue by satra** [nipy/nibabel#1076]
i'm taking an image and simply changing the zoom with this:

`epi_2mm_nii = resample_to_output(epi0, (2,2,2))`

and then plotting the before and after with the niworkflows display module. this results in this image:

 https://dl.dropbox.com/s/tvx6c1wsbatdhgv/test.svg?dl=0 (couldn't figure out a way to display this in github - you will have to download and open with a browser).

i wasn't expecting any visual differences, so now i'm contemplating if this is just an illusion.

**Comment by satra** [comment]
orig affine

```
[[  3.5940001    0.           0.         -65.4303894 ]
 [  0.           3.5940001    0.         -96.33889008]
 [  0.           0.           4.         -86.82946777]
 [  0.           0.           0.           1.        ]]
```

new affine

```
[[  2.           0.           0.         -65.4303894 ]
 [  0.           2.           0.         -96.33889008]
 [  0.           0.           2.         -86.82946777]
 [  0.           0.           0.           1.        ]]
```

**Comment by effigies** [comment]
@satra Did you figure anything out? I think I looked at this back then, but also am not sure if there's an illusion.

**Comment by romainVala** [comment]
Hi
it would say there is an induced shift due to the resampling. It is difficult to be sure because change in resolution also lead to different partial volume, and with high folding of cortical gray matter reslicing may give different shape (for a given slice).

If I compare with mrgrid resampling function of mrtrix I get different result for the affine for instance if I started with 


```
3.600000  0.000000  0.000000  -122.649391
0.000000  3.600000  0.000000  -128.450073
0.000000  0.000000  4.000000  -96.661674
0.000000  0.000000  0.000000  1.000000
```

if I resample to 2 mm iso, resample_to_output lead to the exact same affine matrix (as shown by satra) but mrgrid is giving me a small difference in the transaltion : 

```
 2.000000  0.000000  0.000000  -123.849396
 0.000000  2.000000  0.000000  -129.650070
 0.000000  0.000000  2.000000  -97.661674
 0.000000  0.000000  0.000000  1.000000
```

I am very confident with mrtrix handeling of the affine, but may be it is just due to a different convention. Is it related to  the position of the corner of the voxel  or of the center (as a reference ?... or I am wrong and this is pertinent only for the viewer ... ?) having a shift less than a voxel size is small and may be we do not care, but if one can be exact ...

Since I see difference I compared with a third tool (I tested Resample transform from trochio, which rely on sitk) and .... unfurtunatly I get a third result ...
```
2.000000  0.000000  0.000000  -123.449394
0.000000  2.000000  0.000000  -129.250076
0.000000  0.000000  2.000000  -97.661674
0.000000  0.000000  0.000000  1.000000

```

It also changing the affine (which I think make sense) but end up with a different one ... who's right ???   


I also do this a clear difference of resample_to_output when you have a non strictly axial orientation

For instance  if I start with this affine

```
-1.632297  0.065883  0.102517  114.033714
0.017452  1.494715  -0.652199  -74.317558
0.122627  0.664246  1.457438  -96.661674
0.000000  0.000000  0.000000  1.000000
```

and resample to 1mm iso, nibabel (resample_to_output) give me a strictly axial output

```
1.000000  0.000000  0.000000  -122.649391
0.000000  1.000000  0.000000  -128.450073
0.000000  0.000000  1.000000  -96.661674
0.000000  0.000000  0.000000  1.000000
```

whereas mrtrix gives

```
-0.997133  0.040246  0.064073  114.331818
0.010661  0.913087  -0.407624  -74.570763
0.074910  0.405773  0.910899  -96.905731
0.000000  0.000000  0.000000  1.000000
```

and torchio (sitk) give a similar one

```
 -0.997133  0.040246  0.064073  114.319260
 0.010661  0.913087  -0.407624  -74.489479
 0.074910  0.405773  0.910899  -97.088036
 0.000000  0.000000  0.000000  1.000000
```


For the last example I found nibable output weird:  With a change in the voxel size I expect the global orientation to be unchanged (numerical values of the affine parameters change because of a different total number of voxel, but for both (mrtix and torchio) I get the same global orientation as the input volume

As for the strick axial case, I do see small differences in the translation part between sitk and mrtrix but at least the rotation part are exactly identical !

**Comment by romainVala** [comment]
Yes the small differences I pointed are indeed a matter of the exact chosen reference. This choice should have been define in the format so that every one take the same ... but ok we have to live with imperfect standard definition ...

The important is to stay consistent and also to use the visualization software that goes with the same choice (corner versus center ) 

So we may forget those small differences in the translation part but for the rotation information I would say there is a bug in the nibabel implementation ... it construct a diagonal 3*3 matrix for the affine although it should start from the input affine. 

I would build a zoom affine (identity with zooming factor in the diagonal) and then just divide or multiply with the input affine in order to get the output affine ... 
I hope it make sense

**Comment by mfschmidt** [comment]
I think the edge vs center is probably the case here. I recently needed to build distance matrices with nibabel, and already had a reference from matlab code. When I built distance matrices using nibabel.affines.apply_affine, the cortical surface vertices matched up to matlab's exactly, but the volumes were all off. If I added 1 to every value (2mm isotropic resolution) after apply_affine, the distance matrices were all within float-encoding differences of each other. At the time, I didn't want to assume that matlab was right or nibabel was right. It's all relative. But the matlab version was balanced better by distances to cortical vertices, so I assumed adding one was probably correct in that context.
